README.md
# VNS Script

[![License](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE.md)
[![Version](https://img.shields.io/badge/version-1.0.0-blue.svg)](https://vnscdn.com/)

## Mô tả

VNS Script là một tập lệnh dùng để xây dựng máy chủ chạy web tương thích với Python và Next.js cho hệ điều hành Linux. Tập lệnh này hỗ trợ cài đặt và cập nhật các thành phần như Python, Node.js, Yarn, PostgreSQL, Nginx, Supervisord và các công cụ hữu ích khác.

## Cách sử dụng

Bạn có thể sử dụng các lệnh sau để tải về và chạy VNS Script:

### Sử dụng Curl

```bash
curl -sO https://vnscdn.com/vns.sh && chmod +x vns.sh && bash vns.sh
```

### Sử dụng Wget
  
  ```bash
  wget https://vnscdn.com/vns.sh -O ~/vns.sh && chmod +x ~/vns.sh && bash ~/vns.sh
  ```

## Tùy chọn

- `info`: Hiển thị thông tin script.
- `-h, --help`: Hiển thị hướng dẫn sử dụng.
- `check-update`: Kiểm tra và cập nhật VNS Script và script cài đặt cho hệ điều hành.

## Cron job tự động kiểm tra câp nhật

Script tự động tạo một cron job để kiểm tra cập nhật VNS Script và script cài đặt cho hệ điều hành bằng cách thêm dòng sau vào crontab:

```bash
@reboot bash ~/vns.sh check-update
```

Hoặc bạn có thể tự kiểm tra cập nhật bằng cách chạy thủ công lệnh sau:

```bash
bash ~/vns.sh check-update
```

## Đóng góp và báo lỗi

Nếu bạn gặp lỗi khi sử dụng VNS Script, vui lòng tạo một issue mới tại [đây](https://github.com/nguyenhongthe/VNS/issues) hoặc gửi pull request để đóng góp cho dự án.

## Bản quyền và giấy phép

Copyright (c) 2013-2023 Nguyễn Hồng Thế Blog - Phát hành theo giấy phép [MIT](LICENSE).


changelog.txt
# CHANGELOG VNS SCRIPT - https://vnscdn.com

+ v1.0.2 - 23/09/2023:
    - Sửa lỗi nhỏ.

+ v1.0.1 - 12/09/2023:
    - Thêm hàm quản lý PostgreSQL.

+ v1.0.0 - 25/08/2023:
    - Phát hành lần đầu.
    - Có thể sử dụng được.

vns-db-postgresql.sh
#!/bin/bash
#==============================================================================================================
# Tên tập lệnh: vns-db-postgreslq.sh
# Mô tả: Quản lý cơ sở dữ liệu PostgreSQL.
# Tác giả: Nguyễn Hồng Thế <nguyenhongthe.net>
# Ngày: 2023-09-22
# Phiên bản: 1.0.10
# Giấy phép: Giấy phép MIT
# Sử dụng: curl -sO https://vnscdn.com/vns-db-postgresql.sh && chmod +x vns-db-postgresql.sh && bash vns-db-postgresql.sh
#==============================================================================================================

# PostgreSQL Database Management Script

host="localhost"  # Giá trị mặc định cho host PostgreSQL
port="5432"  # Giá trị mặc định cho cổng PostgreSQL

print_usage() {
    echo "Sử dụng: $0 [Tùy chọn]"
    echo "Tùy chọn:"
    echo "  info, --info            Hiển thị thông tin về script"
    echo "  -h, --help              Hiển thị hướng dẫn sử dụng"
    echo "  -u, --create-user       Tạo người dùng PostgreSQL và cơ sở dữ liệu tương ứng"
    echo "  -d, --delete-user       Xóa người dùng PostgreSQL và cơ sở dữ liệu tương ứng"
    echo "  -e, --empty-schema      Làm rỗng và tạo lại schema"
    echo "  -i, --import-db         Import cơ sở dữ liệu từ file"
    echo "  -x, --export-db         Export cơ sở dữ liệu ra file"
    echo "  -p, --change-password   Đổi mật khẩu cho người dùng postgres"
    exit 0
}

if [[ $# -eq 0 ]]; then
    print_usage
fi

while [[ $# -gt 0 ]]; do
    key="$1"

    case $key in
        -h|--help)
            print_usage
            shift
            ;;
        info|--info)
            info=true
            shift
            ;;
        -u|--create-user)
            create_user=true
            shift
            ;;
        -d|--delete-user)
            delete_user=true
            shift
            ;;
        -e|--empty-schema)
            empty_schema=true
            shift
            ;;
        -i|--import-db)
            import_db=true
            shift
            ;;
        -x|--export-db)
            export_db=true
            shift
            ;;
        -p|--change-password)
            change_postgres_password=true
            shift
            ;;
        *)
            echo "Tùy chọn không hợp lệ: $1"
            print_usage
            exit 1
            ;;
    esac
done

# Thay đổi mật khẩu cho người dùng postgres
if [ "$change_postgres_password" = true ]; then
    echo
    # Nhập tên người dùng cần đổi mật khẩu
    read -p "Nhập tên người dùng cần đổi mật khẩu (mặc định là postgres): " postgres_user
    if [ -z "$postgres_user" ]; then
        postgres_user="postgres"
    fi
    read -s -p "Nhập mật khẩu mới cho người dùng postgres: " postgres_new_password
    sudo -u postgres psql -c "ALTER USER postgres WITH PASSWORD '$postgres_new_password';"
    if [ $? -eq 0 ]; then
        echo "Đổi mật khẩu cho người dùng postgres thành công"
    else
        echo "Lỗi khi đổi mật khẩu cho người dùng postgres"
    fi
fi

# Tạo người dùng và cơ sở dữ liệu mới
if [ "$create_user" = true ]; then
    # Nhập mật khẩu cho người dùng postgres
    echo
    read -s -p "Nhập mật khẩu cho người dùng postgres: " postgres_password
    echo
    read -p "Nhập tên người dùng mới: " new_user
    echo
    read -s -p "Nhập mật khẩu cho người dùng $new_user: " new_user_password
    echo
    read -p "Nhập tên cơ sở dữ liệu mới cho người dùng $new_user: " new_database
    echo

    # Tạo người dùng sử dụng PGPASSWORD
    PGPASSWORD="$postgres_password" psql -U postgres -h "$host" -p "$port" -c "CREATE USER $new_user WITH ENCRYPTED PASSWORD '$new_user_password';"

    # Tạo cơ sở dữ liệu và gán quyền cho người dùng
    PGPASSWORD="$postgres_password" psql -U postgres -h "$host" -p "$port" -c "CREATE DATABASE $new_database OWNER $new_user;"

    # Gán quyền truy cập vào cơ sở dữ liệu
    PGPASSWORD="$postgres_password" psql -U postgres -h "$host" -p "$port" -d "$new_database" -c "GRANT CONNECT ON DATABASE $new_database TO $new_user;"

    # Gán quyền sở hữu vào schema public
    PGPASSWORD="$postgres_password" psql -U postgres -h "$host" -p "$port" -d "$new_database" -c "ALTER SCHEMA public OWNER TO $new_user;"
    PGPASSWORD="$postgres_password" psql -U postgres -h "$host" -p "$port" -d "$new_database" -c "GRANT ALL ON SCHEMA public TO $new_user;"
    unset PGPASSWORD

    if [ $? -eq 0 ]; then
        echo "Tạo người dùng và cơ sở dữ liệu thành công"
    else
        echo "Lỗi khi tạo người dùng và cơ sở dữ liệu"
    fi
fi

# Xóa người dùng và cơ sở dữ liệu
if [ "$delete_user" = true ]; then
    echo
    read -s -p "Nhập mật khẩu PostgreSQL của người dùng quản trị postgres: " postgres_password
    echo
    echo
    echo -e "\e[1;31mBạn đang thực hiện thao tác xóa người dùng và cơ sở dữ liệu PostgreSQL\e[0m"
    echo
    echo -e "\e[1;31mHÃY THAO TÁC CẨN THẬN!\e[0m"
    echo
    read -p "Nhập tên người dùng PostgreSQL cần xóa: " user
    read -p "Nhập tên cơ sở dữ liệu PostgreSQL cần xóa: " database
    echo

    PGPASSWORD="$postgres_password" psql -U postgres -h "$host" -p "$port" -c "DROP DATABASE $database;"
    PGPASSWORD="$postgres_password" psql -U postgres -h "$host" -p "$port" -c "DROP USER $user;"
    unset PGPASSWORD

    if [ $? -eq 0 ]; then
        echo "Xóa người dùng và cơ sở dữ liệu thành công"
    else
        echo "Lỗi khi xóa người dùng và cơ sở dữ liệu"
    fi
fi

# Làm rỗng và tạo lại schema
if [ "$empty_schema" = true ]; then
    read -p "Nhập tên người dùng PostgreSQL: " user
    read -p "Nhập tên cơ sở dữ liệu PostgreSQL: " database
    read -s -p "Nhập mật khẩu PostgreSQL: " password
    echo

    PGPASSWORD="$password" psql -U "$user" -h "$host" -p "$port" -d "$database" -c "DROP SCHEMA public CASCADE;"
    PGPASSWORD="$password" psql -U "$user" -h "$host" -p "$port" -d "$database" -c "CREATE SCHEMA public;"

    if [ $? -eq 0 ]; then
        echo "Xóa và tạo lại schema thành công"
    else
        echo "Lỗi khi xóa hoặc tạo lại schema"
    fi
fi

# Import cơ sở dữ liệu từ file
if [ "$import_db" = true ]; then
    read -p "Nhập tên người dùng PostgreSQL: " user
    read -p "Nhập tên cơ sở dữ liệu PostgreSQL: " database
    read -s -p "Nhập mật khẩu PostgreSQL: " password
    echo
    read -p "Nhập đường dẫn tới file cơ sở dữ liệu để import: " import_file

    PGPASSWORD="$password" pg_restore -U "$user" -h "$host" -p "$port" -d "$database" "$import_file"
    if [ $? -eq 0 ]; then
        echo "Import cơ sở dữ liệu thành công"
    else
        echo "Lỗi khi import cơ sở dữ liệu"
    fi
fi

# Export cơ sở dữ liệu ra file
if [ "$export_db" = true ]; then
    read -p "Nhập tên người dùng PostgreSQL: " user
    read -p "Nhập tên cơ sở dữ liệu PostgreSQL: " database
    read -s -p "Nhập mật khẩu PostgreSQL: " password
    echo
    read -p "Nhập đường dẫn tới file để export cơ sở dữ liệu: " export_file

    PGPASSWORD="$password" pg_dump -U "$user" -h "$host" -p "$port" -Fc "$database" > "$export_file"
    if [ $? -eq 0 ]; then
        echo "Export cơ sở dữ liệu thành công"
    else
        echo "Lỗi khi export cơ sở dữ liệu"
    fi
fi

# Script Info
if [ "$info" = true ]; then
    echo
    echo "================================================================"
    echo "  -   Quản lý cơ sở dữ liệu PostgreSQL."
    echo "  -   Phiên bản: 1.0.10"
    echo "  -   Ngày: 22-09-2023"
    echo "  -   Tác giả: Nguyễn Hồng Thế <nguyenhongthe.net>"
    echo "  -   Giấy phép: MIT License"
    echo "================================================================"
    echo
fi



platforms/vns-web-debian.sh
#!/bin/bash
#=======================================================================================================================
# Tên tập lệnh: vns-web-debian.sh
# Mô tả: Tạo user để chạy web tương thích với Python và Next.js thông qua proxy NGINX trên hệ điều hành Debian.
# Tác giả: Nguyễn Hồng Thế <nguyenhongthe.net>
# Ngày: 2023-09-12
# Phiên bản: 1.0.1
# Giấy phép: Giấy phép MIT
# Sử dụng: curl -sO https://vnscdn.com/platforms/vns-web-debian.sh && chmod +x vns-web-debian.sh && bash vns-web-debian.sh
#=======================================================================================================================

# Hàm tạo thư mục cho user
create_user_directory() {
    local user="$1"
    local directory="$2"

    echo "Tạo thư mục $directory tại /home/$user/$directory thành công. Để lưu trữ các file $directory."
    su - "$user" -c "mkdir -p ~/$directory"
}

# Tạo user
echo "Tạo user"
read -p "Nhập tên user: " username
adduser "$username"

# Chuyển sang user vừa tạo và chạy Zsh
echo "Chuyển sang user vừa tạo và chạy Zsh..."
su - "$username" -c "curl -L http://install.ohmyz.sh | sh && exec zsh -c > /dev/null && chsh -s $(which zsh) > /dev/null"
echo

# Tạo các thư mục cần thiết
echo "Tạo các thư mục cần thiết..."
echo
create_user_directory "$username" "configs"
echo
create_user_directory "$username" "logs"
echo
create_user_directory "$username" "commands"
echo
create_user_directory "$username" "www"
echo
create_user_directory "$username" "nextjs"
echo
create_user_directory "$username" "ssl"
echo
echo "Cấu hình cơ bản cho user $username đã hoàn tất."
echo
# Thoát khỏi user vừa tạo
echo "Thoát khỏi user vừa tạo..."


vns.sh
#!/bin/bash
#==============================================================================================================
# Tên tập lệnh: VNS Script
# Mô tả: Xây dựng máy chủ chạy web tương thích với Python và Next.js cho Linux.
# Tác giả: Nguyễn Hồng Thế <nguyenhongthe.net>
# Ngày: 2023-09-23
# Phiên bản: 1.0.2
# Giấy phép: Giấy phép MIT
# Sử dụng: curl -sO https://vnscdn.com/vns.sh && chmod +x vns.sh && bash vns.sh
# Hoặc: wget https://vnscdn.com/vns.sh -O ~/vns.sh && chmod +x ~/vns.sh && bash ~/vns.sh
#==============================================================================================================

vns_version="1.0.2"
vns_debian_version="1.0.1"
vns_ubuntu_version="1.0.0"
vns_amazon_version="1.0.0"
vns_rhel_version="1.0.0"
vns_website_version="1.0.0"
script_url="https://vnscdn.com/"

# Kiểm tra xem script có được chạy bởi root hay không?
if [ "x$(id -u)" != 'x0' ]; then
    echo 'Bạn cần chạy script này với quyền root'
    exit 1
fi

# Ngó coi thử đang chạy trên hệ điều hành nào.
case $(head -n1 /etc/issue | cut -f 1 -d ' ') in
    Debian)     type="debian" ;;
    Ubuntu)     type="ubuntu" ;;
    Amazon)     type="amazon" ;;
    *)          type="rhel" ;;
esac

# Kiểm tra xem hệ thống đã cài đặt wget hay curl chưa? Nếu chưa thì cài đặt.
if [ ! -e '/usr/bin/wget' ] && [ ! -e '/usr/bin/curl' ]; then
    echo "Cài đặt wget hoặc curl..."
    if [ "$type" == "debian" ] || [ "$type" == "ubuntu" ]; then
        apt-get update -y
        apt-get install -y wget curl
    elif [ "$type" == "amazon" ]; then
        yum update -y
        yum install -y wget curl
    elif [ "$type" == "rhel" ]; then
        dnf update -y
        dnf install -y wget curl
    fi
fi

# Hàm hiển thị hướng dẫn sử dụng.
function display_help() {
    cat <<EOF
Sử dụng: $0 [tùy chọn]

Tùy chọn:
    info            Hiển thị thông tin script.
    -h, --help      Hiển thị hướng dẫn sử dụng.
    check-update    Kiểm tra và cập nhật VNS Script và script cài đặt.
    changelog       In ra changelog của phiên bản mới nhất.
    web             Tạo cấu hình mặc định cho user để chạy web.
    postgresql      Quản lý cơ sở dữ liệu PostgreSQL.
    reinstall       Cài đặt lại VNS Script.
EOF
}

# Kiểm tra số lượng tham số được cung cấp
if [ $# -eq 0 ]; then
    # Kiểm tra xem script đã được cài đặt tại /usr/local/bin/vns hay chưa
    if [ -e "/usr/local/bin/vns" ]; then
        display_help
    else
        # In ra thông tin hệ thống, thuộc hệ điều hành nào.
        echo "Hệ điều hành: $type"
        echo
        # In ra thông tin phiên bản của VNS Script.
        echo "Phiên bản VNS Script: $vns_version"
        echo
        # In ra thông tin phiên bản của VNS Script cho hệ điều hành.
        if [ "$type" == "debian" ]; then
            echo "Phiên bản VNS Script cho Debian: $vns_debian_version"
        elif [ "$type" == "ubuntu" ]; then
            echo "Phiên bản VNS Script cho Ubuntu: $vns_ubuntu_version"
        elif [ "$type" == "amazon" ]; then
            echo "Phiên bản VNS Script cho Amazon Linux: $vns_amazon_version"
        elif [ "$type" == "rhel" ]; then
            echo "Phiên bản VNS Script cho RHEL: $vns_rhel_version"
        fi
        echo
        # In ra thông tin phiên bản của VNS Script cho VNS Website.
        echo "Phiên bản VNS Script cho VNS Website: $vns_website_version"
        echo
        # In ra thông tin phiên bản của các package cần thiết.
        echo "Thông tin phiên bản các package cần thiết:"
        echo
        echo "Python: $python_version"
        echo "Node.js: $node_version"
        echo "Yarn: $yarn_version"
        echo "PostgreSQL: $postgresql_version"
        echo "Nginx: $nginx_version"
        echo "Supervisor: $supervisor_version"
        echo
        # In ra thông tin changelog của phiên bản mới nhất.
        print_changelog
        echo
        # In ra thông tin changelog của phiên bản mới nhất cho hệ điều hành.
        print_platform_changelog
        echo
        # Hỏi người dùng có muốn tải về và cài đặt script cài đặt cho hệ điều hành không?
        read -p "Bạn có muốn tải về và cài đặt script cài đặt cho hệ điều hành $type không? (y/n): " choice
        echo
        # Nếu có thì tải về và cài đặt.
        if [ "$choice" == "y" ]; then
            # Tải về và cài đặt script cài đặt cho hệ điều hành.
            wget "https://vnscdn.com/platforms/vns-$type.sh" -O "$update_script_dir/vns-$type.sh"
            if [ "$?" -eq '0' ]; then
                chmod +x "$update_script_dir/vns-$type.sh"
                bash "$update_script_dir/vns-$type.sh"
                add_update_check_cron
                exit
            else
                echo "Lỗi trong quá trình tải về và cài đặt."
                echo
                exit 1
            fi
        # Nếu không thì bỏ qua.
        else
            echo "Bỏ qua cài đặt script cài đặt cho hệ điều hành $type."
            echo
        fi
    fi
    exit 0
elif [ "$1" == "-h" ] || [ "$1" == "--help" ]; then
    display_help
    exit 0
fi

# Hàm in ra changelog của phiên bản mới của VNS Script (nếu có)
function print_changelog {
    latest_changelog=$(curl -s "https://vnscdn.com/changelog-latest.txt")
    if [ ! -z "$latest_changelog" ]; then
        echo "Changelog của phiên bản mới nhất của VNS Script:"
        echo "$latest_changelog"
        echo "Xem changelog đầy đủ trên trình duyệt: https://vnscdn.com/changelog.txt"
    else
        echo "Không có thông tin changelog cho phiên bản mới."
    fi
}

# Hàm in ra changelog của phiên bản mới của script cài đặt cho hệ điều hành (nếu có)
function print_platform_changelog {
    latest_platform_changelog=$(curl -s "https://vnscdn.com/platforms/changelog-$type-latest.txt")
    if [ ! -z "$latest_platform_changelog" ]; then
        echo "Changelog của phiên bản mới dành cho $type:"
        echo "$latest_platform_changelog"
        echo "Xem changelog đầy đủ trên trình duyệt: https://vnscdn.com/platforms/changelog-$type.txt"
    else
        echo "Không có thông tin changelog cho phiên bản mới của $type."
    fi
}

# Thư mục để chứa script.
update_script_dir=~

# Thêm cron job kiểm tra cập nhật vào crontab.
# Chạy mỗi lần khởi động hệ thống.
function add_update_check_cron {
    echo "Thêm cron job kiểm tra cập nhật vào crontab..."
    echo 
    # Kiểm tra xem cron job đã tồn tại chưa? Nếu chưa thì thêm vào. Nếu có rồi thì té.
    if ! crontab -l | grep -q 'bash ~/vns.sh check-update'; then
        (crontab -l ; echo "@reboot bash ~/vns.sh check-update") | crontab -
    fi
}

# Hàm kiểm tra và tải về script cài đặt cho hệ điều hành.
function check_and_update_platform_script {
    # Kiểm tra xem script cài đặt cho hệ điều hành đã tồn tại chưa? Nếu chưa thì tải về và cài đặt.
    if [ ! -e "$update_script_dir/vns-$type.sh" ]; then
        echo "Tải về script cài đặt cho hệ điều hành $type..."
        echo
        wget "https://vnscdn.com/platforms/vns-$type.sh" -O "$update_script_dir/vns-$type.sh"
        if [ "$?" -eq '0' ]; then
            chmod +x "$update_script_dir/vns-$type.sh"
            bash "$update_script_dir/vns-$type.sh" "$*"
            add_update_check_cron
            exit
        else
            echo "Lỗi trong quá trình tải về và cài đặt."
            echo
            exit 1
        fi
    # Ngược lại thì kiểm tra xem phiên bản hiện tại của script cài đặt cho hệ điều hành có khác phiên bản mới nhất hay không?
    else
        # Lấy phiên bản mới nhất của script cài đặt cho hệ điều hành từ repo của VNSCDN.
        latest_version_platform=$(curl -s "https://vnscdn.com/platforms/latest-version-$type.txt")
        # Lấy phiên bản hiện tại của script cài đặt cho hệ điều hành.
        current_version_platform=$(grep -oE "vns_${type}_version=\"([0-9.]+)\"" "$update_script_dir/vns-$type.sh" | cut -d '"' -f 2)
        echo "Script cài đặt cho hệ điều hành $type đã được cài đặt."
        echo
        echo "Phiên bản hiện tại của script cài đặt cho hệ điều hành là $type: $current_version_platform"
        echo
        # Kiểm tra xem phiên bản hiện tại của script cài đặt cho hệ điều hành có khác phiên bản mới nhất hay không?
        if [ "$latest_version_platform" != "$current_version_platform" ]; then
            echo "Phiên bản mới của script cài đặt cho hệ điều hành $type là: $latest_version_platform"
            echo
            # In ra changelog của phiên bản của hệ điều hành.
            print_platform_changelog
            echo
            # Nếu khác thì hỏi người dùng có muốn tải về và cập nhật không?
            read -p "Bạn có muốn tải về và cập nhật phiên bản $latest_version_platform cho $type không? (y/n): " choice
            echo
            # Nếu có thì tải về và cập nhật.
            if [ "$choice" == "y" ]; then
                wget "https://vnscdn.com/platforms/vns-$type.sh" -O "$update_script_dir/vns-$type.sh"
                # Kiểm tra xem tải về và cập nhật thành công hay không?
                if [ "$?" -eq '0' ]; then
                    chmod +x "$update_script_dir/vns-$type.sh"
                    bash "$update_script_dir/vns-$type.sh" "$*"
                    echo
                    echo "Cập nhật phiên bản cho vns-$type.sh hoàn tất."
                    echo
                    exit
                # Nếu không thì báo lỗi và thoát.
                else
                    echo "Lỗi trong quá trình tải về và cập nhật vns-$type.sh."
                    echo
                    exit 1
                fi
            # Nếu không thì bỏ qua.
            else
                echo "Bỏ qua cập nhật vns-$type.sh."
                echo
            fi
        # Nếu không thì báo là phiên bản mới nhất.
        else
            echo "Script vns-$type.sh đã là phiên bản mới nhất."
            echo
        fi
    fi
}

# Hàm kiểm tra và cập nhật cả VNS Script và script cài đặt cho hệ điều hành.
function check_update {
    echo "Kiểm tra cập nhật phiên bản VNS Script..."
    echo
    # Kiểm tra xem VNS Script đã tồn tại chưa? Nếu chưa thì tải về và cài đặt.
    if [ ! -e "$update_script_dir/vns.sh" ]; then
        echo "Tải về VNS Script..."
        echo
        wget "https://vnscdn.com/vns.sh" -O "$update_script_dir/vns.sh"
        if [ "$?" -eq '0' ]; then
            chmod +x "$update_script_dir/vns.sh"
            bash "$update_script_dir/vns.sh" "$*"
            add_update_check_cron
            exit
        else
            echo "Lỗi trong quá trình tải về và cài đặt VNS Script."
            echo
            exit 1
        fi
    # Nếu có rồi thì in ra thông báo đã cài đặt và phiên bản hiện tại.
    else
        # Lấy phiên bản mới nhất của VNS Script từ repo của VNSCDN.
        latest_version=$(curl -s "https://vnscdn.com/latest-version.txt")
        echo "VNS Script đã được cài đặt."
        echo
        echo "Phiên bản hiện tại của VNS Script là: $vns_version"
        echo
        # Kiểm tra xem phiên bản hiện tại của VNS Script có khác phiên bản mới nhất hay không?
        if [ "$latest_version" != "$vns_version" ]; then
            echo "Mhiên bản mới của VNS Script là: $latest_version"
            echo
            # In ra changelog của phiên bản mới của VNS Script.
            print_changelog
            echo
            # Nếu khác thì hỏi người dùng có muốn tải về và cập nhật không?
            read -p "Bạn có muốn tải về và cập nhật phiên bản của VNS Script không? (y/n): " choice
            echo
            # Nếu có thì tải về và cập nhật.
            if [ "$choice" == "y" ]; then
                wget "https://vnscdn.com/vns.sh" -O "$update_script_dir/vns.sh"
                # Kiểm tra xem tải về và cập nhật thành công hay không?
                if [ "$?" -eq '0' ]; then
                    chmod +x "$update_script_dir/vns.sh"
                    bash "$update_script_dir/vns.sh" "$*"
                    echo
                    echo "Cập nhật phiên bản VNS Script hoàn tất. "
                    echo
                    exit
                # Nếu không thì báo lỗi và thoát.
                else
                    echo "Lỗi trong quá trình tải về và cập nhật phiên bản VNS Script."
                    echo
                    exit 1
                fi
            # Nếu không thì bỏ qua.
            else
                echo "Bỏ qua cập nhật phiên bản VNS Script."
                echo
            fi
        # Nếu không thì báo là phiên bản mới nhất.
        else
            echo "VNS Script hiện tại đã là phiên bản mới nhất."
            echo
        fi
    fi

    # Gọi hàm kiểm tra và cập nhật script cài đặt cho hệ điều hành.
    check_and_update_platform_script "$*"
}

# Hàm tạo web cho user với cấu hình mặc định.
function vns_web {
    # Kiểm tra xem script tạo web đã tồn tại chưa? Nếu chưa thì tải script tương ứng cho hệ điều hành từ repo của VNSCDN về và chạy nó.
    if [ ! -e "$update_script_dir/vns-web-$type.sh" ]; then
        wget "https://vnscdn.com/platforms/vns-web-$type.sh" -O "$update_script_dir/vns-web-$type.sh"
        if [ "$?" -eq '0' ]; then
            chmod +x "$update_script_dir/vns-web-$type.sh"
            bash "$update_script_dir/vns-web-$type.sh" "$*"
            echo
            echo "Cài đặt web cho user hoàn tất."
            echo
            exit
        else
            echo "Lỗi trong quá trình tải về và cài đặt web cho user."
            echo
            exit 1
        fi
    # Nếu có rồi thì chạy luôn khỏi cần tải.
    else
        bash "$update_script_dir/vns-web-$type.sh" "$*"
        echo
        echo "Cài đặt web cho user hoàn tất."
        echo
        exit
    fi
}

# Hàm quản lý cơ sở dữ liệu postgresql.
function vns_db_postgresql {
    # Kiểm tra xem script quản lý cơ sở dữ liệu postgresql đã tồn tại chưa? Nếu chưa thì tải về và chạy nó.
    if [ ! -e "$update_script_dir/vns-db-postgresql.sh" ]; then
        wget "https://vnscdn.com/vns-db-postgresql.sh" -O "$update_script_dir/vns-db-postgresql.sh"
        if [ "$?" -eq '0' ]; then
            chmod +x "$update_script_dir/vns-db-postgresql.sh"
            bash "$update_script_dir/vns-db-postgresql.sh"
            echo
            echo "Cài đặt script quản lý cơ sở dữ liệu PostgreSQL hoàn tất."
            echo
            exit
        else
            echo "Lỗi trong quá trình tải về và cài đặt script quản lý cơ sở dữ liệu PostgreSQL."
            echo
            exit 1
        fi
    # Nếu có rồi thì chạy luôn khỏi cần tải.
    else
        bash "$update_script_dir/vns-db-postgresql.sh"
        echo
        echo "Cài đặt script quản lý cơ sở dữ liệu PostgreSQL hoàn tất."
        echo
        exit
    fi
}

# Lấy phiên bản Python nếu có hoặc thông báo nếu chưa cài đặt.
python_version=$(python3 -V 2>&1)
if [ $? -eq 0 ]; then
    python_version="Phiên bản Python: $python_version"
else
    python_version="Python chưa được cài đặt."
fi

# Lấy phiên bản Node.js nếu có hoặc thông báo nếu chưa cài đặt.
node_version=$(node -v 2>&1)
if [ $? -eq 0 ]; then
    node_version="Phiên bản Node.js: $node_version"
else
    node_version="Node.js chưa được cài đặt."
fi

# Lấy phiên bản Yarn nếu có hoặc thông báo nếu chưa cài đặt.
yarn_version=$(yarn -v 2>&1)
if [ $? -eq 0 ]; then
    yarn_version="Phiên bản Yarn: $yarn_version"
else
    yarn_version="Yarn chưa được cài đặt."
fi

# Lấy phiên bản PostgreSQL nếu có hoặc thông báo nếu chưa cài đặt.
postgresql_version=$(psql --version 2>&1)
if [ $? -eq 0 ]; then
    postgresql_version="Phiên bản PostgreSQL: $postgresql_version"
else
    postgresql_version="PostgreSQL chưa được cài đặt."
fi

# Lấy phiên bản Nginx nếu có hoặc thông báo nếu chưa cài đặt.
nginx_version=$(nginx -v 2>&1 | grep -o '[0-9.]*')
if [ $? -eq 0 ]; then
    nginx_version="Phiên bản Nginx: $nginx_version"
else
    nginx_version="Nginx chưa được cài đặt."
fi

# Lấy phiên bản Supervisor nếu có hoặc thông báo nếu chưa cài đặt.
supervisor_version=$(supervisord -v 2>&1)
if [ $? -eq 0 ]; then
    supervisor_version="Phiên bản Supervisor: $supervisor_version"
else
    supervisor_version="Supervisor chưa được cài đặt."
fi

# Hiển thị thông tin tác giả khi sử dụng tùy chọn 'info'.
if [ "$1" == "info" ]; then
    cat <<EOF
$0 
#=====================================================================================================
    Tên tập lệnh: VNS Script                                                                               
    Mô tả: Xây dựng máy chủ chạy web tương thích với Python và Next.js cho Linux.                                                                       
    Tác giả: Nguyễn Hồng Thế <nguyenhongthe.net>                                                        
    Ngày: 2023-09-23                                                                                   
    Phiên bản: $vns_version                                                                                    
    Giấy phép: Giấy phép MIT                                                                            
    Sử dụng: curl -sO https://vnscdn.com/vns.sh && chmod +x vns.sh && bash vns.sh
=====================================================================================================
=====================================================================================================
                                    THÔNG TIN PHIÊN BẢN PACKAGE                                      
=====================================================================================================
=====================================================================================================
    - Phiên bản VNS Script: $vns_version                                                                
    - Phiên bản VNS Script cho Debian: $vns_debian_version
    - Phiên bản VNS Script cho Ubuntu: $vns_ubuntu_version
    - Phiên bản VNS Script cho Amazon Linux: $vns_amazon_version
    - Phiên bản VNS Script cho RHEL: $vns_rhel_version
    - Phiên bản VNS Script cho VNS Website: $vns_website_version                                         
    - Script url: $script_url                                                                           
    - $python_version
    - $node_version
    - $yarn_version
    - $postgresql_version
    - $nginx_version
    - $supervisor_version
=====================================================================================================
=====================================================================================================
    Mẹo: Nếu bạn muốn thêm VNS Script vào biến môi trường PATH thì có thể chạy lệnh sau:
    "ln -s $update_script_dir/vns.sh /usr/local/bin/vns"
    Sau đó bạn có thể chạy VNS Script bằng cách gõ lệnh 'vns' ở bất kỳ thư mục nào.
=====================================================================================================

EOF
    exit 0
fi

# Kiểm tra nếu tham số là 'web' thì chỉ chạy hàm tạo web.
if [ "$1" == "web" ]; then
    vns_web
    exit
fi

# Kiểm tra nếu tham số là 'postgresql' thì chỉ chạy hàm quản lý cơ sở dữ liệu postgresql.
if [ "$1" == "postgresql" ]; then
    vns_db_postgresql
    exit
fi

# Kiểm tra nếu tham số là 'changelog' thì chạy hàm in ra changelog của phiên bản mới nhất.
if [ "$1" == "changelog" ]; then
    print_changelog
    echo
    print_platform_changelog
    exit
fi

# Kiểm tra nếu tham số là 'reinstall' thì tải mới về phiên bản mới nhất và chạy lại hàm cài đặt VNS Script.
if [ "$1" == "reinstall" ]; then
    update_script_dir=~
    curl -sO https://vnscdn.com/vns.sh
    bash "$update_script_dir/vns.sh"
    exit
fi

# Kiểm tra nếu tham số là 'reinstall-$type' thì tải mới về phiên bản mới nhất và chạy hàm cài đặt cho $type.
if [ "$1" == "reinstall-$type" ]; then
    update_script_dir=~
    curl -sO "https://vnscdn.com/platforms/vns-$type.sh"
    bash "$update_script_dir/vns-$type.sh"
    exit
fi 

# Kiểm tra nếu tham số là 'check-update' thì chạy hàm kiểm tra và cập nhật VNS Script và script cài đặt cho hệ điều hành.
if [ "$1" == "check-update" ]; then
    check_update
    exit
fi

exit
